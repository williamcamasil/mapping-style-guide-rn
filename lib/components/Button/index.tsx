import React, { ComponentType } from 'react';
import {
  StyleProp, StyleSheet, TouchableOpacity, TouchableWithoutFeedbackProps, View, ViewStyle,
} from 'react-native';

import { SvgProps } from 'react-native-svg';

import { useViewStyles } from '../../hooks/useStyles';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import Loading from '../Loading';
import Spacer from '../Spacer';
import Text, { TextPropTypes } from '../Text';
import {
  ButtonSizeType,
  ButtonVariantType,
  getSizeStyle,
  getTextColor,
  getVariantStyle,
  getStylesToBlock,
} from './utils';

export type ButtonPropsType<IconProps extends SvgProps = SvgProps> = {
  children?: TextPropTypes['children'];
  onPress: TouchableWithoutFeedbackProps['onPress'];
  style?: StyleProp<ViewStyle>;
  disabled?: TouchableWithoutFeedbackProps['disabled'];
  theme: AppThemeType;
  variant?: ButtonVariantType;
  size?: ButtonSizeType;
  loading?: boolean;
  testID?: string;
  Icon?: ComponentType<IconProps>;
  block?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
  textColorVariant?: ButtonVariantType;
};

const styles = StyleSheet.create({
  containerIconAndText: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

const Button: React.FC<ButtonPropsType> = ({
  children,
  style,
  onPress,
  disabled,
  theme,
  variant,
  size,
  loading,
  testID,
  Icon,
  block,
  containerStyle,
  textColorVariant,
}) => {
  const hasText = Boolean(children);

  const buttonStyles = useViewStyles(
    () => [
      {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: theme.borders.radius.medium,
      },
      getVariantStyle(theme, variant, disabled),
      getSizeStyle(theme, size, hasText),
      style,
    ],
    [disabled, hasText, size, style, theme, variant],
  );

  const containerStyles = useViewStyles(() => [getStylesToBlock(block), containerStyle], [block, containerStyle]);

  const textColor = getTextColor(textColorVariant ? textColorVariant : variant, disabled);

  const renderText = () => {
    if (!hasText) return null;
    return (
      <Text
        testID="text-with-icon"
        variant={size === 'small' ? 'bodySmall' : 'body'}
        weight="bold"
        lineHeight="medium"
        color={textColor}
      >
        {children}
      </Text>
    );
  };

  const renderContent = () => {
    if (loading) {
      return <Loading color={textColor} />;
    }

    if (Icon) {
      return (
        <View testID="container-icon" style={styles.containerIconAndText}>
          <Icon width={32} height={32} color={theme.colors[textColor]} />
          {children ? <Spacer size={theme.spacings.sXXS} /> : null}
          {renderText()}
        </View>
      );
    }

    return renderText();
  };

  return (
    <View testID="container-button" style={containerStyles}>
      <TouchableOpacity
        testID={testID}
        style={buttonStyles}
        onPress={onPress}
        disabled={disabled || loading}
        activeOpacity={theme.opacities.intense}
      >
        {renderContent()}
      </TouchableOpacity>
    </View>
  );
};

Button.defaultProps = {
  disabled: false,
  variant: 'containedPrimary',
  size: 'medium',
  block: true,
};

export default withTheme(Button);
