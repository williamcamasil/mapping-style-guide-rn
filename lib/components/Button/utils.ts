import { ViewStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type ButtonVariantType =
  'containedPrimary' |
  'containedSecondary' |
  'outlinedPrimary' |
  'outlinedSecondary' |
  'text';

export type ButtonSizeType =
  'large' |
  'medium' |
  'small';

export const getVariantStyle = (
  theme: AppThemeType,
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ViewStyle => {
  switch (variant) {
    case 'containedSecondary':
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primary200,
        borderWidth: theme.borders.width.none,
      };
    case 'outlinedPrimary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.primary100 : theme.colors.primary600,
      };
    case 'outlinedSecondary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.neutralGray100 : theme.colors.neutralGray300,
      };
    case 'text':
      return {
        backgroundColor: disabled ? theme.colors.neutralGray100 : 'transparent',
        borderWidth: theme.borders.width.none,
      };
    case 'containedPrimary':
    default:
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primaryMain,
        borderWidth: theme.borders.width.none,
      };
  }
};

export const getTextColor = (
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ColorPaletesType => {
  switch (variant) {
    case 'containedSecondary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedPrimary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedSecondary':
      return disabled ? 'neutralGray200' : 'neutralGray600';
    case 'text':
      return disabled ? 'neutralGray400' : 'neutralGray500';
    case 'containedPrimary':
    default:
      return disabled ? 'primary300' : 'neutralWhite';
  }
};

export const getSizeStyle = (theme: AppThemeType, size?: ButtonSizeType, hasText?: boolean): ViewStyle => {
  switch (size) {
    case 'large':
      return {
        height: 64,
        minWidth: 64,
        paddingHorizontal: hasText ? theme.spacings.sXL : undefined,
      };
    case 'small':
      return {
        height: 30,
        minWidth: 30,
        paddingHorizontal: hasText ? theme.spacings.sXS : undefined,
      };
    case 'medium':
    default:
      return {
        height: 48,
        minWidth: 48,
        paddingHorizontal: hasText ? theme.spacings.sMedium : undefined,
      };
  }
};

export const getStylesToBlock = (block: boolean | undefined): ViewStyle | null => {
  if (block) return null;

  return {
    flexDirection: 'row',
  };
};
