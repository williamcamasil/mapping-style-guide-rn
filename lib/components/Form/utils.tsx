import type { FormApi, MutableState } from 'final-form';

export const hasValue = (value: any) => {
  if (value === undefined) return false;
  if (value === null) return false;

  if (Array.isArray(value)) return value.length > 0;

  if (value instanceof Date) return true;

  switch (typeof value) {
    case 'object':
      return Object.keys(value).length > 0;
    case 'number':
    case 'boolean':
      return true;
    case 'string':
    default:
      return Boolean(value);
  }
};

export function getFieldErrorState(
  invalid: boolean | undefined,
  touched: boolean | undefined,
  error?: any,
  submitError?: any,
  helpMessage?: string,
) {
  const fieldError = error || submitError;

  const showError = Boolean(invalid && touched);
  const message = showError && typeof fieldError === 'string'
    ? fieldError
    : helpMessage;

  return {
    showError,
    message,
  };
}

type SetTouchedArguments = Parameters<(name: string, touched: boolean) => void>;

export function setFormFieldTouched<T>(form: FormApi<T>, fieldName: string, touched: boolean) {
  /* istanbul ignore next */
  if (!form.mutators.setFieldTouched) {
    console.warn('O form n√£o possui o mutator setFieldTouched');
    return;
  }
  form.mutators.setFieldTouched(fieldName, touched);
}

export function setFieldTouched<FormValues, InitialFormValues = Partial<FormValues>>(
  args: SetTouchedArguments,
  state: MutableState<FormValues, InitialFormValues>,
) {
  const [name, touched] = args;
  const field = state.fields[name];
  if (field) {
    field.touched = Boolean(touched);
  }
}

