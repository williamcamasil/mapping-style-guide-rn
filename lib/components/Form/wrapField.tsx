import React, { useCallback } from 'react';

import { FieldState } from 'final-form';
import { Field, FieldMetaState, FieldRenderProps } from 'react-final-form';

import { hasValue } from './utils';

export type WrapFieldProps<FieldValue> = FieldMetaState<FieldValue> & {
  value?: FieldValue | null;
  onChange?: (value: FieldValue | null) => void;
  onBlur?: (event?: any) => void;
  onFocus?: (event?: any) => void;
  required?: boolean;
};

export type FieldValidator<FieldValue, FormValues = any, InitialFormValues = Partial<FormValues>> = (
  value: FieldValue,
  allValues?: InitialFormValues,
  meta?: FieldState<FieldValue>
) => any | Promise<any>;

export type WrappedFieldRenderProps<FieldValue> = {
  /**
   * Nome do field que será retornado no objeto `values` ao submeter o formulário.
   */
  name: string;
  /**
   * Função que valida o estado do field.
   * Se o field está válido, então deve retornar `undefined`.
   * Caso contrário, deve retornar a `string` com a mensagem de erro.
   */
  validate?: FieldValidator<FieldValue>;
  /**
   * Função que faz o parse do valor informado antes de salvar no `state` do
   * formulário:
   *
   * field -> parse -> formState
   */
  parse?: (value: FieldValue, name: string) => FieldValue;
  /**
   * Função que formata o valor do campo para renderizar na tela:
   *
   * formState -> format -> field
   */
  format?: (value: FieldValue, name: string) => FieldValue;
  /**
   * Quando requerido, o `validate` é aplicado mesmo com o field vazio.
   * Se não for requerido, então o `validate` é aplicado apenas quando o field tiver um valor.
   */
  required?: boolean;
  /**
   * valor inicial de um campo
  */
  initialValue?: FieldValue;
};

export default function wrapField<
  Props,
  FieldValue = string | undefined,
  ExternalProps = Omit<Props, keyof WrapFieldProps<FieldValue>>,
>(renderField: (componentProps: ExternalProps, fieldProps: WrapFieldProps<FieldValue>) => React.ReactElement) {

  type FieldWrapperPropsType = WrappedFieldRenderProps<FieldValue> & ExternalProps;

  function FieldWrapper({
    name,
    validate,
    parse,
    format,
    required,
    initialValue,
    ...othersFieldProps
  }: FieldWrapperPropsType) {
    const renderFieldProps = useCallback((renderProps: FieldRenderProps<FieldValue>) => renderField(
      /**
       * Foi necessário desativar passar any aqui devido a uma limitação de tipagem do FinalForm.
       * Assim é possível inferir o tipo corretamente ao componente retornado pelo "wrapField".
       */
      (othersFieldProps as any),
      {
        ...renderProps.input,
        ...renderProps.meta,
        required,
      },
    ), [othersFieldProps, required]);

    const handleValidation: FieldValidator<FieldValue> = useCallback((value, values, meta) => {
      if (validate && (required || hasValue(value))) {
        return validate(value, values, meta);
      }
      return undefined;
    }, [validate, required]);

    return (
      <Field<FieldValue>
        name={name}
        render={renderFieldProps}
        validate={handleValidation}
        parse={parse}
        format={format}
        initialValue={initialValue}
      />
    );
  }

  return FieldWrapper;
}
