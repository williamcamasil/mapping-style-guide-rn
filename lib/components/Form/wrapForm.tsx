import React, {
  ComponentType, useCallback, useRef,
} from 'react';

import {
  FormApi, Mutator, SubmissionErrors, ValidationErrors,
} from 'final-form';
import { delay } from 'mapping-context-rn';
import { Form, FormRenderProps } from 'react-final-form';

import { setFieldTouched } from './utils';

export type OnSubmitFormType<FormValues, InitialFormValues = Partial<FormValues>> = (
  values: FormValues,
  form: FormApi<FormValues, InitialFormValues>,
) => SubmissionErrors | Promise<SubmissionErrors | void> | void;

type WrapFormRenderProps<FormValues, InitialFormValues> = Omit<FormRenderProps<FormValues, InitialFormValues>, 'handleSubmit'> & {
  handleSubmit: (
    submitCallback: OnSubmitFormType<FormValues, InitialFormValues>
  ) => () => void;
};

type WrapFormOptionsType<FormValues, InitialFormValues> = {
  form?: FormApi<FormValues, InitialFormValues>;
  initialValues?: InitialFormValues;
  validate?: (
    values: FormValues,
  ) => ValidationErrors | Promise<ValidationErrors>;
  validateOnBlur?: boolean;
  mutators?: { [key: string]: Mutator<FormValues, InitialFormValues> };
};

export default function wrapForm<
  Props,
  FormValues,
  InitialFormValues = Partial<FormValues>,
>(
  FormComponent: ComponentType<Props & WrapFormRenderProps<FormValues, InitialFormValues>>,
  options: WrapFormOptionsType<FormValues, InitialFormValues> = {},
) {

  type FormWrapperPropsType = Props & {
    form?: FormApi<FormValues, InitialFormValues>;
  };

  const defaultOptions: WrapFormOptionsType<FormValues, InitialFormValues> = {
    ...options,
    mutators: {
      setFieldTouched,
      ...options.mutators,
    },
  };

  function FormWrapper({ form: injectedForm, ...formWrapperProps }: FormWrapperPropsType) {
    const handleSubmitRef = useRef<FormRenderProps<FormValues, InitialFormValues>['handleSubmit']>();
    const submitCallbackRef = useRef<OnSubmitFormType<FormValues, InitialFormValues>>();

    const handleSubmitCreator = useCallback((submitCallback: OnSubmitFormType<FormValues, InitialFormValues>) => {
      submitCallbackRef.current = submitCallback;

      return () => handleSubmitRef.current?.();
    }, []);

    const handleFormSubmit = useCallback(async (
      values: FormValues,
      form: FormApi<FormValues, InitialFormValues>,
    ) => {
      // aguarda 1ms para que o form possa atualizar seu estado interno antes de submeter
      await delay(1);
      return submitCallbackRef.current?.(values, form);
    }, []);

    const renderForm = useCallback(({
      handleSubmit,
      ...othersRenderProps
    }: FormRenderProps<FormValues, InitialFormValues>) => {
      /**
       * Armazena a função de submit que irá disparar a submissão do formulário.
       */
      handleSubmitRef.current = handleSubmit;

      return (
        <FormComponent
          {...formWrapperProps as Props}
          {...othersRenderProps}
          /**
           * Passa para o componente a função que irá "criar" a função de submissão.
           */
          handleSubmit={handleSubmitCreator}
        />
      );
    }, [formWrapperProps, handleSubmitCreator]);

    return (
      <Form<FormValues, InitialFormValues>
        {...defaultOptions}
        onSubmit={handleFormSubmit}
        form={injectedForm ?? defaultOptions?.form}
      >
        {renderForm}
      </Form>
    );
  }

  FormWrapper.displayName = `FormWrapper(${FormComponent.displayName})`;

  return FormWrapper;
}
