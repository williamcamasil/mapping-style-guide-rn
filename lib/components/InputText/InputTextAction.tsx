import React, { ComponentType } from 'react';
import {
  ColorValue, StyleSheet, TouchableOpacity,
} from 'react-native';

import { SvgProps } from 'react-native-svg';

import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { INPUT_TEXT_HEIGHT } from './utils';

export type InputTextActionPropType<IconProps extends SvgProps = SvgProps> = {
  theme: AppThemeType;
  onPress: () => void;
  Icon: ComponentType<IconProps>;
  color?: ColorValue;
  size: number;
  testID?: string;
  disabled?: boolean;
};

const styled = StyleSheet.create({
  icon: {
    position: 'absolute',
    justifyContent: 'center',
    alignItems: 'center',
    right: 0,
    top: 0,
    height: INPUT_TEXT_HEIGHT,
    width: INPUT_TEXT_HEIGHT,
  },
});

const InputTextAction: React.FC<InputTextActionPropType> = ({
  theme,
  onPress: handlePress,
  Icon,
  size,
  color,
  testID,
  disabled,
}) => (
  <TouchableOpacity
    testID={testID}
    style={styled.icon}
    onPress={handlePress}
    activeOpacity={theme.opacities.intense}
    disabled={disabled}
  >
    <Icon width={size} height={size} color={color || theme.colors.neutralGray400} />
  </TouchableOpacity>
);

InputTextAction.defaultProps = {
  size: 24,
  testID: 'input-text-action',
};

export default withTheme(InputTextAction);
