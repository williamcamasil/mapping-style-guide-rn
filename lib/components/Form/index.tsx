import React from 'react';

import { FormApi, Mutator, ValidationErrors } from 'final-form';
import { Form as FinalForm, FormRenderProps } from 'react-final-form';

import { OnSubmitFormType } from './wrapForm';

export { default as wrapForm } from './wrapForm';
export { default as wrapField } from './wrapField';
export { hasValue, getFieldErrorState, setFormFieldTouched } from './utils';

export type { OnSubmitFormType } from './wrapForm';
export type {
  WrappedFieldRenderProps, WrapFieldProps, FieldValidator,
} from './wrapField';

export type FormPropsType<FormValues, InitialFormValues> = {
  children: (props: FormRenderProps<FormValues, InitialFormValues>) => React.ReactNode;
  onSubmit: OnSubmitFormType<FormValues, InitialFormValues>;
  form?: FormApi<FormValues, InitialFormValues>;
  initialValues?: InitialFormValues;
  validate?: (
    values: FormValues,
  ) => ValidationErrors | Promise<ValidationErrors>;
  validateOnBlur?: boolean;
  mutators?: { [key: string]: Mutator<FormValues, InitialFormValues> };
};

function Form<FormValues, InitialFormValues = Partial<FormValues>>(
  props: FormPropsType<FormValues, InitialFormValues>,
): JSX.Element {
  return (
    <FinalForm<FormValues, InitialFormValues> {...props} />
  );
}

export default Form;
