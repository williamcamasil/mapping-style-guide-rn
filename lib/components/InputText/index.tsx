import React, { useCallback, useState } from 'react';
import {
  NativeSyntheticEvent,
  StyleProp,
  TextInput as RNTextInput,
  TextInputFocusEventData,
  TextInputProps,
  View,
  ViewStyle,
} from 'react-native';

import { useTextStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { getFieldErrorState, wrapField } from '../Form';
import HelpMessage from '../HelpMessage';
import InputTextAction, { InputTextActionPropType } from './InputTextAction';
import InputTextLabel from './InputTextLabel';
import ShowHiddenPasswordButton from './ShowHiddenPasswordButton';
import {
  getErrorStyle,
  getTextStyle,
  getViewStyle,
  getActionStyle,
  INPUT_TEXT_HEIGHT,
} from './utils';

export type InputTextPropsType = Omit<TextInputProps, 'placeholder' | 'onChange' | 'onChangeText' | 'editable' | 'pointerEvents'> & {
  theme: AppThemeType;
  label: string;
  containerStyle?: StyleProp<ViewStyle>;
  value: string | null;
  onChange?: TextInputProps['onChangeText'];
  help?: string;
  error?: boolean;
  disabled?: boolean;
  /**
   * Desabilita o Input sem alterar o estilo.
   */
  readOnly?: boolean;
  action?: React.ReactElement<InputTextActionPropType, typeof InputTextAction>;
  inputPointerEvents?: TextInputProps['pointerEvents'];
  innerRef?: React.LegacyRef<RNTextInput>;
  textArea?: boolean;
};

const InputText = ({
  theme,
  style,
  label,
  value,
  containerStyle,
  onChange,
  onFocus,
  onBlur,
  help,
  error,
  disabled,
  readOnly,
  secureTextEntry,
  action,
  inputPointerEvents,
  innerRef,
  textArea,
  ...others
}: InputTextPropsType): React.ReactElement => {
  const [focused, setFocused] = useState(false);

  const [showPassword, setShowPassword] = useState(false);

  const hasValue = Boolean(value);
  const hasAction = Boolean(action ?? secureTextEntry);

  const inputStyles = useTextStyles(
    () => [
      getTextStyle(theme, focused, hasValue, disabled),
      getViewStyle(theme, INPUT_TEXT_HEIGHT, focused, hasValue, textArea, disabled),
      getErrorStyle(theme, error),
      getActionStyle(hasAction),
      style,
    ],
    [theme, focused, hasValue, disabled, textArea, error, hasAction, style],
  );

  const handleFocus = useCallback(
    (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
      setFocused(true);
      onFocus?.(e);
    },
    [onFocus],
  );

  const handleBlur = useCallback(
    (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
      setFocused(false);
      onBlur?.(e);
    },
    [onBlur],
  );

  const handleShowOrHidden = useCallback((show: boolean) => {
    setShowPassword(show);
  }, []);

  const renderAction = () => {
    if (action) {
      return action;
    }

    return (
      <ShowHiddenPasswordButton
        visible={secureTextEntry}
        showPassword={showPassword}
        onShowPasswordChange={handleShowOrHidden}
        disabled={disabled}
      />
    );
  };

  return (
    <View style={containerStyle}>
      <RNTextInput
        testID="rn-text-input"
        autoCapitalize={secureTextEntry ? 'none' : undefined}
        autoComplete={secureTextEntry ? 'off' : undefined}
        autoCorrect={secureTextEntry ? false : undefined}
        spellCheck={secureTextEntry ? false : undefined}
        {...others}
        ref={innerRef}
        secureTextEntry={Boolean(secureTextEntry && !showPassword)}
        editable={Boolean(!disabled && !readOnly)}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onChangeText={onChange}
        value={value}
        placeholderTextColor={theme.colors.neutralGray500}
        placeholder={label}
        style={inputStyles}
        pointerEvents={inputPointerEvents}
      />
      <HelpMessage
        error={error}
        disabled={disabled}
        marginLeft={!textArea ? theme.spacings.sSmall : theme.spacings.sZero}
      >
        {help}
      </HelpMessage>

      <InputTextLabel hide={!hasValue} focused={focused} error={error} textArea={textArea}>
        {label}
      </InputTextLabel>

      {renderAction()}
    </View>
  );
};

InputText.defaultProps = {
  allowFontScaling: false,
};

InputText.Field = withTheme(wrapField<InputTextPropsType>((props, fieldProps) => {
  const {
    invalid, touched, error, submitError, value, onChange, onBlur, onFocus,
  } = fieldProps;

  const { showError, message } = getFieldErrorState(
    invalid,
    touched,
    error,
    submitError,
    props.help,
  );

  return (
    <InputText
      {...props}
      value={value || ''}
      onChange={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
      error={showError}
      help={message}
    />
  );
}));

InputText.Action = InputTextAction;

export default withTheme(InputText);
