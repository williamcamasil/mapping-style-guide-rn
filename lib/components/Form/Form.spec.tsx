import React from 'react';
import { View } from 'react-native';

import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { createForm } from 'final-form';
import { useDidMount } from 'mapping-context-rn';

import Form from '.';
import Button from '../Button';
import InputText from '../InputText';
import {
  getFieldErrorState, hasValue, setFieldTouched, setFormFieldTouched,
} from './utils';
import wrapForm from './wrapForm';

describe('InputText.Field', () => {
  it('snapshot', () => {
    const result = render(
      <Form
        onSubmit={jest.fn()}
      >
        {() => (
          <InputText.Field
            name="inputName"
            label="Input label"
            help="Error message"
          />
        )}
      </Form>,
    );

    const tree = result.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('invalid', () => {
    const form = createForm({
      onSubmit: jest.fn(),
      initialValues: {
        inputName: '',
      },
    });

    render(
      <Form
        onSubmit={jest.fn()}
        form={form}
      >
        {() => (
          <InputText.Field
            name="inputName"
            label="Input label"
            required
            validate={() => 'Erro'}
          />
        )}
      </Form>,
    );

    expect(form.getFieldState('inputName')?.invalid).toBeTruthy();
  });

  it('active', () => {
    const form = createForm({
      onSubmit: jest.fn(),
      initialValues: {
        inputName: '',
      },
    });

    const result = render(
      <Form
        onSubmit={jest.fn()}
        form={form}
      >
        {() => (
          <InputText.Field
            name="inputName"
            label="Input label"
            help="Error message"
            required
            validate={() => 'Erro'}
          />
        )}
      </Form>,
    );

    fireEvent(result.getByTestId('rn-text-input'), 'focus');

    expect(form.getFieldState('inputName')?.active).toBeTruthy();

    fireEvent(result.getByTestId('rn-text-input'), 'blur');

    expect(form.getFieldState('inputName')?.touched).toBeTruthy();
  });

  it('wrapped', async () => {
    const handleSubmitPress = jest.fn();

    const MyScreen = wrapForm(({ handleSubmit, form }) => {
      useDidMount(() => {
        form.initialize({
          inputName: 'foo-bar',
        });
      });

      return (
        <>
          <InputText.Field
            testID="form-input-text"
            name="inputName"
            label="Input label"
            required
            validate={() => undefined}
          />
          <Button
            testID="form-button"
            onPress={handleSubmit(values => handleSubmitPress(values))}
          >
            Enviar
          </Button>
        </>
      );
    });

    const result = render(
      <MyScreen />,
    );

    await result.findByDisplayValue('foo-bar');

    fireEvent.press(result.getByTestId('form-button'));

    await waitFor(() => {
      expect(handleSubmitPress).toBeCalledWith({ inputName: 'foo-bar' });
    });
  });

  it('wrapped optional invalid', () => {
    const form = createForm({
      onSubmit: jest.fn(),
    });

    const MyScreen = wrapForm(
      () => (
        <InputText.Field
          name="inputName"
          label="Input label"
          required={false}
          validate={value => (value !== 'foo-bar' ? 'error' : undefined)}
        />
      ),
      {
        form,
      },
    );

    const result = render(
      <MyScreen />,
    );

    expect(form.getState().valid).toBeTruthy();

    const textInput = result.getByTestId('rn-text-input');

    fireEvent(textInput, 'focus');
    fireEvent(textInput, 'changeText', 'abcd');
    fireEvent(textInput, 'blur');

    expect(form.getState().invalid).toBeTruthy();
  });

  it('Should set the inputName2 field as touched', async () => {
    const MyScreen = wrapForm(
      ({ form }) => {
        const handleSetTouched = () => {
          setFormFieldTouched(form, 'inputName2', true);
        };

        return (
          <View>
            <InputText.Field
              name="inputName"
              label="Input label"
              required
              validate={value => (value !== 'foo-bar' ? 'Input label is touched' : undefined)}
            />
            <InputText.Field
              name="inputName2"
              label="Input label 2"
              required
              validate={value => (value !== 'foo-bar' ? 'Input label 2 is touched' : undefined)}
            />
            <Button testID="set-touched-button" onPress={handleSetTouched}>Test</Button>
          </View>
        );
      },
    );

    const result = render(
      <MyScreen />,
    );

    expect(result.queryByText('Input label is touched')).toBeNull();
    expect(result.queryByText('Input label 2 is touched')).toBeNull();

    fireEvent.press(result.getByTestId('set-touched-button'));

    await result.findByText('Input label 2 is touched');
    expect(result.queryByText('Input label is touched')).toBeNull();
  });

  it('Should merge the wrapForm options, customMutator and setFieldTouched should exist', async () => {
    const MyScreen = wrapForm(
      ({ form }) => {
        const handleSetTouched = () => {
          form.mutators.customMutator();
          setFormFieldTouched(form, 'inputName2', true);
        };

        return (
          <View>
            <InputText.Field
              name="inputName"
              label="Input label"
              required
              validate={value => (value !== 'foo-bar' ? 'Input label is touched' : undefined)}
            />
            <InputText.Field
              name="inputName2"
              label="Input label 2"
              required
              validate={value => (value !== 'foo-bar' ? 'Input label 2 is touched' : undefined)}
            />
            <Button testID="set-touched-button" onPress={handleSetTouched}>Test</Button>
          </View>
        );
      },
      {
        mutators: {
          customMutator: () => {},
        },
      },
    );

    const result = render(
      <MyScreen />,
    );

    expect(result.queryByText('Input label 2 is touched')).toBeNull();

    fireEvent.press(result.getByTestId('set-touched-button'));

    await result.findByText('Input label 2 is touched');
  });

  it('wrapped submit', async () => {
    const handleFormSubmit = jest.fn();

    const MyScreen = wrapForm(({ handleSubmit }) => (
      <>
        <InputText.Field
          name="inputName"
          label="Input label"
        />
        <Button testID="button-submit" onPress={handleSubmit(handleFormSubmit)}>
          Enviar
        </Button>
      </>
    ));

    const result = render(
      <MyScreen />,
    );

    fireEvent.press(result.getByTestId('button-submit'));

    await waitFor(() => {
      expect(handleFormSubmit).toBeCalled();
    });
  });

});

describe('Utils', () => {
  it('hasValue', () => {
    expect(hasValue(undefined)).toBeFalsy();
    expect(hasValue(null)).toBeFalsy();
    expect(hasValue('')).toBeFalsy();
    expect(hasValue({})).toBeFalsy();
    expect(hasValue([])).toBeFalsy();

    expect(hasValue(0)).toBeTruthy();
    expect(hasValue(1)).toBeTruthy();
    expect(hasValue('a')).toBeTruthy();
    expect(hasValue({
      key: 'key',
      label: 'label',
    })).toBeTruthy();
    expect(hasValue(new Date())).toBeTruthy();
    expect(hasValue(new Date(0))).toBeTruthy();
    expect(hasValue(false)).toBeTruthy();
    expect(hasValue(true)).toBeTruthy();
  });

  it('getFieldErrorState showError', () => {
    expect(getFieldErrorState(true, false).showError).toBeFalsy();
    expect(getFieldErrorState(false, false).showError).toBeFalsy();
    expect(getFieldErrorState(false, true).showError).toBeFalsy();
    expect(getFieldErrorState(true, true).showError).toBeTruthy();
  });

  it('getFieldErrorState help message', () => {
    expect(getFieldErrorState(false, false, 'Validation error', 'Submission error', 'Help message')).toEqual({
      showError: false,
      message: 'Help message',
    });
  });

  it('getFieldErrorState validation error', () => {
    expect(getFieldErrorState(true, true, 'Validation error', undefined, 'Help message')).toEqual({
      showError: true,
      message: 'Validation error',
    });
  });

  it('getFieldErrorState submission error', () => {
    expect(getFieldErrorState(true, true, undefined, 'Submission error', 'Help message')).toEqual({
      showError: true,
      message: 'Submission error',
    });
  });

  it('getFieldErrorState validation and submission error', () => {
    expect(getFieldErrorState(true, true, 'Validation error', 'Submission error', 'Help message')).toEqual({
      showError: true,
      message: 'Validation error',
    });
  });

  it('should set field from states as touched', () => {
    const field1: any = {};
    const fields: any = { field1 };
    const state: any = { fields };
    setFieldTouched(['field1', true], state);
    expect(field1.touched).toBe(true);
  });

  it('should set field from states as not touched', () => {
    const field1: any = { touched: true };
    const fields: any = { field1 };
    const state: any = { fields };
    setFieldTouched(['field1', false], state);
    expect(field1.touched).toBe(false);
  });

  it('should not set field from states as touched, field is undefined', () => {
    const field1: any = {};
    const fields: any = { field1 };
    const state: any = { fields };
    setFieldTouched(['field2', true], state);
    expect(fields.field2?.touched).toBeUndefined();
  });
});
