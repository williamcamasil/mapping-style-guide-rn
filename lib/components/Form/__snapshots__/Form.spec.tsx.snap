// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText.Field snapshot 1`] = `
<View>
  <TextInput
    allowFontScaling={false}
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Input label"
    placeholderTextColor="#A4A7AF"
    secureTextEntry={false}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E1E2E4",
        "borderRadius": 12,
        "borderWidth": 1,
        "color": "#676D79",
        "fontFamily": "Averta CY",
        "fontSize": 16,
        "fontWeight": "normal",
        "height": 64,
        "letterSpacing": -0.02,
        "lineHeight": 16,
        "paddingBottom": 0,
        "paddingLeft": 20,
        "paddingTop": 4,
      }
    }
    testID="rn-text-input"
    value=""
  />
  <Text
    allowFontScaling={false}
    style={
      {
        "color": "#676D79",
        "flexShrink": 1,
        "fontFamily": "Averta CY",
        "fontSize": 14,
        "fontWeight": "normal",
        "letterSpacing": -0.02,
        "lineHeight": 14,
        "marginLeft": 20,
        "marginTop": 12,
      }
    }
  >
    Error message
  </Text>
</View>
`;
